### PR TITLE
[enhancement](Nereids) support otherJoinConjuncts in cascades join reorder

### PR DESCRIPTION
**Pull request by jackwener**
# Proposed changes

Issue Number: close #xxx

## Problem summary

support otherJoinConjuncts in cascades join reorder.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [x] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [x] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [x] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 38.94 seconds
 load time: 558 seconds
 storage size: 17154655744 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221026145522_clickbench_pr_34459.html

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/join/InnerJoinLAsscomProject.java [hunk 1]
```diff
@@ -111,14 +117,26 @@ public Rule build() {
...
-                    // TODO: Does nonHashOnCondition also need to be considered.
-                    Map<Boolean, Set<Slot>> abOnUsedSlots = newTopHashJoinConjuncts.stream()
+                    // replace otherJoinConjuncts
+                    newBottomOtherJoinConjuncts = JoinUtils.replaceJoinConjuncts(
+                            newBottomOtherJoinConjuncts, outputToInput);
+                    newTopOtherJoinConjuncts = JoinUtils.replaceJoinConjuncts(
+                            newTopOtherJoinConjuncts, inputToOutput);
+
+                    if (newBottomHashJoinConjuncts == null || newTopHashJoinConjuncts == null
+                            || newBottomOtherJoinConjuncts == null || newTopOtherJoinConjuncts == null) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/join/InnerJoinLAsscomProject.java:132]: why prohibit other join conjuncts is null

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/join/InnerJoinLAsscomProject.java [hunk 2]
```diff
@@ -172,15 +192,17 @@ public static Set<ExprId> getExprIdSetForB(GroupPlan b, List<NamedExpression> bP
...
     /**
-     * Split HashCondition into two part.
+     * Split Condition into two part.
+     * True: contains B.
+     * False: just contains A C.
      */
-    public static Map<Boolean, List<Expression>> splitHashConjunctsWithAlias(List<Expression> topHashConjuncts,
-            LogicalJoin<GroupPlan, GroupPlan> bottomJoin, Set<ExprId> bExprIdSet) {
+    public static Map<Boolean, List<Expression>> splitConjunctsWithAlias(List<Expression> topConjuncts,
+            LogicalJoin<GroupPlan, GroupPlan> bottomJoin, List<Expression> bottomConjunct, Set<ExprId> bExprIdSet) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/join/InnerJoinLAsscomProject.java:200]:
```suggestion
            LogicalJoin<GroupPlan, GroupPlan> bottomJoin, List<Expression> bottomConjuncts, Set<ExprId> bExprIdSet) {
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/join/OuterJoinLAsscomProject.java [hunk 3]
```diff
@@ -89,20 +87,36 @@ public Rule build() {
...
-                    List<Expression> newTopHashJoinConjuncts = splitOn.get(true);
+                    /* ********** split Conjuncts ********** */
+                    Map<Boolean, List<Expression>> splitHashJoinConjuncts
+                            = InnerJoinLAsscomProject.splitConjunctsWithAlias(
+                            topJoin.getHashJoinConjuncts(), bottomJoin, bottomJoin.getHashJoinConjuncts(), bExprIdSet);
+                    List<Expression> newTopHashJoinConjuncts = splitHashJoinConjuncts.get(true);
+                    Preconditions.checkState(!newTopHashJoinConjuncts.isEmpty(),
+                            "LAsscom newTopHashJoinConjuncts join can't empty");
+                    // When newTopHashJoinConjuncts.size() != bottomJoin.getHashJoinConjuncts().size()
+                    // It means that topHashJoinConjuncts contain A, B, C, we should LAsscom.
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/join/OuterJoinLAsscomProject.java:98]:
```suggestion
                    // It means that topHashJoinConjuncts contain A, B, C, we should not do LAsscom.
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/join/OuterJoinLAsscomProject.java [hunk 4]
```diff
@@ -117,14 +131,25 @@ public Rule build() {
...
                             newTopHashJoinConjuncts, inputToOutput);
+                    // replace otherJoinConjuncts
+                    newBottomOtherJoinConjuncts = JoinUtils.replaceJoinConjuncts(
+                            newBottomOtherJoinConjuncts, outputToInput);
+                    newTopOtherJoinConjuncts = JoinUtils.replaceJoinConjuncts(
+                            newTopOtherJoinConjuncts, inputToOutput);
+
+                    if (newBottomHashJoinConjuncts == null || newTopHashJoinConjuncts == null
+                            || newBottomOtherJoinConjuncts == null || newTopOtherJoinConjuncts == null) {
+                        return null;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/join/OuterJoinLAsscomProject.java:146]: other conjuncts could be null, and it is ok i think

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/join/OuterJoinLAsscomProject.java [hunk 5]
```diff
@@ -89,20 +87,36 @@ public Rule build() {
...
                     }
-                    List<Expression> newBottomHashJoinConjuncts = splitOn.get(false);
+                    List<Expression> newBottomHashJoinConjuncts = splitHashJoinConjuncts.get(false);
                     if (newBottomHashJoinConjuncts.size() == 0) {
                         return null;
                     }
 
-                    /* ********** replace HashConjuncts by projects ********** */
+                    Map<Boolean, List<Expression>> splitOtherJoinConjuncts
+                            = InnerJoinLAsscomProject.splitConjunctsWithAlias(
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/join/OuterJoinLAsscomProject.java:109]: if splitConjunctsWithAlias used both for Inner and Outer, it should be move to JoinUtils

**Review by morrySnow — APPROVED** [review]

**Pull request merged**